File: packages/python/qraudio/codec/envelope.py
from __future__ import annotations

import math
from typing import MutableSequence


def applyFade(samples: MutableSequence[float], sample_rate: float, fade_ms: float) -> None:
    fade_samples = max(0, round((fade_ms / 1000.0) * sample_rate))
    if fade_samples == 0 or fade_samples * 2 > len(samples):
        return
    for i in range(fade_samples):
        t = i / fade_samples
        gain = 0.5 * (1 - math.cos(math.pi * t))
        samples[i] *= gain
        samples[len(samples) - 1 - i] *= gain

File: packages/js/src/core/codec/envelope.ts
export function applyFade(samples: Float32Array, sampleRate: number, fadeMs: number): void {
  const fadeSamples = Math.max(0, Math.round((fadeMs / 1000) * sampleRate));
  if (fadeSamples === 0 || fadeSamples * 2 > samples.length) {
    return;
  }
  for (let i = 0; i < fadeSamples; i += 1) {
    const t = i / fadeSamples;
    const gain = 0.5 * (1 - Math.cos(Math.PI * t));
    samples[i] *= gain;
    samples[samples.length - 1 - i] *= gain;
  }
}
